keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ResourceService.kt

```diff
@@ -0,0 +1,192 @@
+    // without explicit action from the user to resume.
+    if (actuationPauser.isPaused(id)) {
+      return PAUSED
+    }
+
+    val history = getEnrichedEventHistory(id, 10)
+    return when {
+      history.isHappy() -> HAPPY
+      history.isUnhappy() -> UNHAPPY
+      history.isDiff() -> DIFF
+      history.isActuating() -> ACTUATING
+      history.isError() -> ERROR
+      history.isCreated() -> CREATED
+      history.isVetoed() -> VETOED
+      history.isResumed() -> RESUMED
+      history.isCurrentlyUnresolvable() -> CURRENTLY_UNRESOLVABLE
+      else -> UNKNOWN
+    }
+  }
+
+  private fun List<PersistentEvent>.isHappy(): Boolean {
+    return first() is ResourceValid || first() is ResourceDeltaResolved
+  }
+
+  private fun List<PersistentEvent>.isActuating(): Boolean {
+    return first() is ResourceActuationLaunched || first() is ResourceTaskSucceeded ||
+      // we might want to move ResourceTaskFailed to isError later on
+      first() is ResourceTaskFailed
+  }
+
+  private fun List<PersistentEvent>.isError(): Boolean {
+    return first() is ResourceCheckError
+  }
+
+  private fun List<PersistentEvent>.isCreated(): Boolean {
+    return first() is ResourceCreated
+  }
+
+  private fun List<PersistentEvent>.isDiff(): Boolean {
+    return first() is ResourceDeltaDetected || first() is ResourceMissing
+  }
+
+  private fun List<PersistentEvent>.isPaused(): Boolean {
+    val appPaused = firstOrNull { it is ApplicationActuationPaused }?.timestamp
+    val appResumed = firstOrNull { it is ApplicationActuationResumed }?.timestamp
+
+    return first() is ResourceActuationPaused ||
+      // If the app was paused and has not yet resumed
+      (appPaused != null && (appResumed == null || appResumed.isBefore(appPaused)))
+  }
+
+  private fun List<PersistentEvent>.isVetoed(): Boolean {
+    return first() is ResourceActuationVetoed
+  }
+
+  private fun List<PersistentEvent>.isResumed(): Boolean {
+    return first() is ResourceActuationResumed || first() is ApplicationActuationResumed
+  }
+
+  private fun List<PersistentEvent>.isCurrentlyUnresolvable(): Boolean {
+    return first() is ResourceCheckUnresolvable
+  }
+
+  /**
+   * Checks last 10 events for flapping between only ResourceActuationLaunched and ResourceDeltaDetected
+   */
+  private fun List<PersistentEvent>.isUnhappy(): Boolean {
+    val recentSliceOfHistory = this.subList(0, Math.min(10, this.size))
+    val filteredHistory = recentSliceOfHistory.filter { it is ResourceDeltaDetected || it is ResourceActuationLaunched }
+    if (filteredHistory.size != recentSliceOfHistory.size) {
+      // there are other events, we're not thrashing.
+      return false
+    }
+    return true
+  }
+
+  /**
+   * Returns a list of [ResourceSummary] for the specified application.
+   *
+   * This function assumes there's a single delivery config associated with the application.
+   */
```

Comment:
I don't agree that this goes here, but I have a question on the changed implementation. Before this had a sql implementation, now it does not. Is that because you have all the information needed in the delivery config, and we actually didn't need to look in the database?

Reply:
Correct. It was redundant with the `ResourceRepository.get()` method.

Reply:
Correct.